src/kernels/ipywidgets-message-coordination/cdnWidgetScriptSourceProvider.ts

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+'use strict';
+
+import { ConfigurationTarget, Memento } from 'vscode';
+import { IApplicationShell } from '../../platform/common/application/types';
+import { Telemetry } from '../../platform/common/constants';
+import {
+    IConfigurationService,
+    IHttpClient,
+    WidgetCDNs
+} from '../../platform/common/types';
+import { createDeferred, Deferred } from '../../platform/common/utils/async';
+import { Common, DataScience } from '../../platform/common/utils/localize';
+import { noop } from '../../platform/common/utils/misc';
+import { traceError, traceInfo, traceVerbose } from '../../platform/logging';
+import { ConsoleForegroundColors } from '../../platform/logging/types';
+import { sendTelemetryEvent } from '../../telemetry';
+import { IWidgetScriptSourceProvider, WidgetScriptSource } from './types';
+
+// Source borrowed from https://github.com/jupyter-widgets/ipywidgets/blob/54941b7a4b54036d089652d91b39f937bde6b6cd/packages/html-manager/src/libembed-amd.ts#L33
+const unpgkUrl = 'https://unpkg.com/';
+const jsdelivrUrl = 'https://cdn.jsdelivr.net/npm/';
+
+const GlobalStateKeyToTrackIfUserConfiguredCDNAtLeastOnce = 'IPYWidgetCDNConfigured';
+const GlobalStateKeyToNeverWarnAboutScriptsNotFoundOnCDN = 'IPYWidgetNotFoundOnCDN';
+
+function moduleNameToCDNUrl(cdn: string, moduleName: string, moduleVersion: string) {
```

Comment:
We don't need two seprate clases now.
When dealing with CDN, we cannot save the CDN file to disk, that breaks widgets.
If user doesn't have access to the internet we'll display  a warning message indicating the fact that cdn isn't being used due to lack of connectivity.

Reply:
This could be a rename instead of an add? Then we wouldn't lose history.